File: GaoKaoNotes/dataStructure/programming/cLanguageTest/test68.py
def heaps_algorithm(n, arr, result):
    if n == 1:
        result.append(arr.copy())
        return
    # Actually, the whole logic is by locking the last elements and then swap around the previous ones recursively. the two rules of heap's algorithm guarantees that the permutations won't repeat itself. It doesn't like normal swap methods and used[] methods would backtrack back to orginal state. Heap's algorithm would use permutation to generate another different permutation. You can prove this is true by "mathmematical induction proof", prove that that small n, the advantages of heap's algorithm are true. And this can derive that for big n, it also be true.
    for i in range(n):
        heaps_algorithm(n - 1, arr, result)

        # Swap depending on whether n is odd or even, but the truth is odd and even is not important at all, the only things matter is that you have to switch the bewteen the two rules when n changes
        if n % 2 == 1:
            arr[0], arr[n - 1] = arr[n - 1], arr[0]
        else:
            arr[i], arr[n - 1] = arr[n - 1], arr[i]

        ''' This below also works perfectly which swithces the rules between even and odds. The algorithm itself is symmetry between two rules.
        if n % 2 == 0:
            arr[0], arr[n - 1] = arr[n - 1], arr[0]
        else:
            arr[i], arr[n - 1] = arr[n - 1], arr[i]
        '''
        # the only things matters the logic should be consistent through all the process


# Example usage
data = [1, 2, 3, 4, 5]
result = []
heaps_algorithm(len(data), data, result)

print("Permutations:")
for perm in result:
    print(perm)

File: GaoKaoNotes/dataStructure/programming/cLanguageTest/test68.c
// heap's algorithm, c language version
#include <stdio.h>
#include <stdlib.h>

#define arr_length 4

void swap(int *a, int *b) {
    int temp = *a;
    *a = *b;
    *b = temp;
    return;
}

void permute(int nums[arr_length], int n) {
    if (n==1) {
        for (int i=0; i<arr_length; i++) printf("%d->", nums[i]);
        printf("\n");
        return;
    }
       
    for (int i=0; i<n; i++) {
        permute(nums, n-1);
        // the 2 rules below doesn't matter if n is odd or even unless it would switch when n changes
        if (n%2) swap(&nums[0], &nums[n-1]);
        else swap(&nums[0], &nums[n-1]);
    }
    return;
}

int main() {
    int nums[arr_length] = {1, 2, 3, 4};
    permute(nums, arr_length);
    return 0;
}
